source/extensions/common/wasm/context.cc

```diff
@@ -0,0 +1,212 @@
+  } else if (type == MetricType::Gauge) {
+    auto id = wasm()->nextGaugeMetricId();
+    auto g = &wasm()->scope_->gaugeFromStatName(stat_name, Stats::Gauge::ImportMode::Accumulate);
+    wasm()->gauges_.emplace(id, g);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  } else if (type == MetricType::Histogram) {
+    auto id = wasm()->nextHistogramMetricId();
+    auto h = &wasm()->scope_->histogramFromStatName(stat_name, Stats::Histogram::Unit::Unspecified);
+    wasm()->histograms_.emplace(id, h);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  }
+  return WasmResult::BadArgument;
+}
+
+WasmResult Context::incrementMetric(uint32_t metric_id, int64_t offset) {
+  auto type = static_cast<MetricType>(metric_id & Wasm::kMetricTypeMask);
+  if (type == MetricType::Counter) {
+    auto it = wasm()->counters_.find(metric_id);
+    if (it != wasm()->counters_.end()) {
+      if (offset > 0) {
+        it->second->add(offset);
+        return WasmResult::Ok;
+      } else {
+        return WasmResult::BadArgument;
+      }
+      return WasmResult::NotFound;
+    }
+  } else if (type == MetricType::Gauge) {
+    auto it = wasm()->gauges_.find(metric_id);
+    if (it != wasm()->gauges_.end()) {
+      if (offset > 0) {
+        it->second->add(offset);
+        return WasmResult::Ok;
+      } else {
+        it->second->sub(-offset);
+        return WasmResult::Ok;
+      }
+    }
+    return WasmResult::NotFound;
+  }
+  return WasmResult::BadArgument;
+}
+
+WasmResult Context::recordMetric(uint32_t metric_id, uint64_t value) {
+  auto type = static_cast<MetricType>(metric_id & Wasm::kMetricTypeMask);
+  if (type == MetricType::Counter) {
+    auto it = wasm()->counters_.find(metric_id);
+    if (it != wasm()->counters_.end()) {
+      it->second->add(value);
+      return WasmResult::Ok;
+    }
+  } else if (type == MetricType::Gauge) {
+    auto it = wasm()->gauges_.find(metric_id);
+    if (it != wasm()->gauges_.end()) {
+      it->second->set(value);
+      return WasmResult::Ok;
+    }
+  } else if (type == MetricType::Histogram) {
+    auto it = wasm()->histograms_.find(metric_id);
+    if (it != wasm()->histograms_.end()) {
+      it->second->recordValue(value);
+      return WasmResult::Ok;
+    }
+  }
+  return WasmResult::NotFound;
+}
+
+WasmResult Context::getMetric(uint32_t metric_id, uint64_t* result_uint64_ptr) {
+  auto type = static_cast<MetricType>(metric_id & Wasm::kMetricTypeMask);
+  if (type == MetricType::Counter) {
+    auto it = wasm()->counters_.find(metric_id);
+    if (it != wasm()->counters_.end()) {
+      *result_uint64_ptr = it->second->value();
+      return WasmResult::Ok;
+    }
+    return WasmResult::NotFound;
+  } else if (type == MetricType::Gauge) {
+    auto it = wasm()->gauges_.find(metric_id);
+    if (it != wasm()->gauges_.end()) {
+      *result_uint64_ptr = it->second->value();
+      return WasmResult::Ok;
+    }
+    return WasmResult::NotFound;
+  }
+  return WasmResult::BadArgument;
```

Comment:
Is it clear that histograms are not supported in this API? Link to ABI definition please?

Reply:
This is a limitation of Envoy.  There is no get functionality for histograms AFAICT.

Reply:
Histograms work when allocated from ThreadLocalStore. They do not work when allocated from IsolatedStoreimpl.

Reply:
I am just using context.scope().createScope("") where context is the filter context.  What might I substitute to give me a ThreadLocalStore and what are the tradeoffs with using one?  I was under the impression that we wanted to use a sub-Scope in general for filters.

Reply:
There are generally two types of stores: IsolatedStoreImpl and ThreadLocalStore. The main store used in production is ThreadLocalStore, but some subsystems create isolated stores, e.g.hosts.

Histograms should just work as long as you are not using an isolated store, and their data should be visible in the admin /stats output and to sinks, eg Prometheus.

Reply:
The WASM code should definitely be using the system's ThreadLocalStore, this will take care of histograms, scopes, etc. properly. (Note I haven't reviewed this in depth yet because I'm still waiting on doing the ABI review.)

Reply:
Great, because that is what it is using.